packages/flutter_tools/lib/src/web/chrome.dart

```diff
@@ -0,0 +1,177 @@
+      '--disable-background-timer-throttling',
+      // Since we are using a temp profile, disable features that slow the
+      // Chrome launch.
+      '--disable-extensions',
+      '--disable-popup-blocking',
+      '--bwsi',
+      '--no-first-run',
+      '--no-default-browser-check',
+      '--disable-default-apps',
+      '--disable-translate',
+      url,
+    ];
+    final Process process = await processManager.start(args);
+
+    // Wait until the DevTools are listening before trying to connect.
+    await process.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .firstWhere((String line) => line.startsWith('DevTools listening'))
+        .timeout(const Duration(seconds: 60), onTimeout: () {
+          throwToolExit('Unable to connect to Chrome DevTools.');
+          return null;
+        });
+
+    return _connect(Chrome._(
+      port,
+      ChromeConnection('localhost', port),
+      process: process,
+      dataDir: dataDir,
+    ));
+  }
+
+  static Future<Chrome> _connect(Chrome chrome) async {
+    if (_currentCompleter.isCompleted) {
+      throwToolExit('Only one instance of chrome can be started.');
+    }
+    // The connection is lazy. Try a simple call to make sure the provided
+    // connection is valid.
+    try {
+      await chrome.chromeConnection.getTabs();
+    } catch (e) {
+      await chrome.close();
+      throwToolExit(
+          'Unable to connect to Chrome debug port: ${chrome.debugPort}\n $e');
+    }
+    _currentCompleter.complete(chrome);
+    return chrome;
+  }
+
+  /// Connects to an instance of Chrome with an open debug port.
+  static Future<Chrome> fromExisting(int port) async =>
+      _connect(Chrome._(port, ChromeConnection('localhost', port)));
+
+  static Future<Chrome> get connectedInstance => _currentCompleter.future;
+}
+
+/// A class for managing an instance of Chrome.
+class Chrome {
+  const Chrome._(
+    this.debugPort,
+    this.chromeConnection, {
+    Process process,
+    Directory dataDir,
+  })  : _process = process,
+        _dataDir = dataDir;
+
+  final int debugPort;
+  final Process _process;
+  final Directory _dataDir;
+  final ChromeConnection chromeConnection;
+
+  static Completer<Chrome> _currentCompleter = Completer<Chrome>();
+
+  Future<void> close() async {
+    if (_currentCompleter.isCompleted) {
+      _currentCompleter = Completer<Chrome>();
+    }
+    chromeConnection.close();
+    _process?.kill(ProcessSignal.SIGKILL);
+    await _process?.exitCode;
+    try {
+      // Chrome starts another process as soon as it dies that modifies the
+      // profile information. Give it some time before attempting to delete
+      // the directory.
+      await Future<void>.delayed(const Duration(milliseconds: 500));
+    } catch (_) {
+      // Silently fail if we can't clean up the profile information.
+    } finally {
+      await _dataDir?.delete(recursive: true);
```

Comment:
Probably also need to not die if this throws a FileSystemException if the delete fails.